### PR TITLE
Data source for the Analytics App

### PR DESCRIPTION
**Pull request by Rello**
This pull adds a data source to report on forms with Analyitcs

- [x] real time reporting
- [x] scheduled data loads (for trends)
- [x] charts and/or tables
- [x] multi page dashboards (coming with Analytics 5) 
- [x] PDF/print dashboards (coming with Analytics 5) 
- [ ] one-click dashboard creation for all questions of a form

---

### Data source selection
Bildschirmfoto 2024-06-01 um 17 43 25

---

### Single report
Bildschirmfoto 2024-06-01 um 17 46 31

---

### Dashboard
Bildschirmfoto 2024-06-01 um 17 48 57

### CODE REVIEW

**Comment by susnux** [comment]
CI is failing a lot (PHP code style, missing stubs for psalm, missing unit tests), please fix this :)

Also cc @Chartman123, what do you think about this?

**Comment by Chartman123** [comment]
@susnux if this doesn't come with any problems when the Analytics app isn't installed, I'd be fine with it.

**Comment by Rello** [comment]
@Chartman123 I just did another test on my side - all clean. no nextcloud.log
happy for an other retest

...we only need to find a way for the psalm and the unit-tests...

File: psalm.xml [hunk 1]
```diff
@@ -35,10 +37,12 @@
 				<referencedClass name="Doctrine\DBAL\Schema\SchemaException" />
 				<referencedClass name="Doctrine\DBAL\Driver\Statement" />
 				<referencedClass name="Doctrine\DBAL\Schema\Table" />
+				<referencedClass name="OCA\Analytics\Datasource\DatasourceEvent" />
 			</errorLevel>
 		</UndefinedDocblockClass>
 	</issueHandlers>
 	<stubs>
 		<file name="tests/stubs/oc_hooks_emitter.php" />
+		<file name="tests/stubs/OCA-Analytics-IDatasource.php" />
```
Comment (Fenn-CS) [inline psalm.xml:46]: We want to maintain the naming convention here?

```suggestion
		<file name="tests/stubs/oca-analytics-idatasource.php" />
```

and hence file rename?

**Review by Fenn-CS — APPROVED** [review]
Look okay.

**Comment by Chartman123** [comment]
@Rello are you ready with this PR? Perhaps you could add some more unit tests, so that the Codecov workflows turn green, too 🙂

Before we merge this, also please do a rebase on current main and squash all the commits into one single commit.

**Comment by Rello** [comment]
Hi @Chartman123 
I will retest one more time and post an update here. 
regarding unit tests, I would need assistant. I never did these in the past as the dev-capacity is pretty limited...

**Comment by Chartman123** [comment]
@Rello please don't merge the main branch into your branch and use `git rebase` instead :)

**Comment by Rello** [comment]
> @Rello please don't merge the main branch into your branch and use `git rebase` instead :)

Hi,
any hint on the background? I was always doing this until now. but open to learn something better.
Tx!

**Comment by Chartman123** [comment]
Rebasing results in a cleaner commit history, when we merge the changes back into our main :)

**Comment by Rello** [comment]
@Chartman123 @susnux 
all tests are green - besides the codecov part and I have no idea how to solve this.
what do you suggest to do here?

File: lib/Analytics/AnalyticsDatasource.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * Analytics data source
+ * Report Table App data with the Analytics App
```
Comment (susnux) [inline lib/Analytics/AnalyticsDatasource.php:4]: copy paste issue?

**Review by susnux — APPROVED** [review]
ok for me

File: lib/Analytics/AnalyticsDatasource.php (hunk 2 above)
Comment (Rello) [reply lib/Analytics/AnalyticsDatasource.php]: oh, yes. thank you. corrected

**Comment by Rello** [comment]
🙏🏻

**Comment by Chartman123** [comment]
@susnux we should have squashed the commits into one single commit first... ;)

**Pull request merged**